lib/saluki-components/src/destinations/datadog_metrics/endpoint/endpoints.rs

```diff
@@ -0,0 +1,152 @@
+use std::{
+    collections::{HashMap, HashSet},
+    str::FromStr,
+};
+
+use http::Uri;
+use regex::Regex;
+use saluki_error::GenericError;
+use saluki_metadata;
+use serde::Deserialize;
+use snafu::{ResultExt, Snafu};
+use url::Url;
+
+use crate::destinations::datadog_metrics::DEFAULT_SITE;
+
+/// AdditionalEndpoints holds the endpoints and api keys parsed
+/// from the user configuration.
+#[derive(Clone, Debug, Default, Deserialize)]
+pub struct AdditionalEndpoints(HashMap<String, Vec<String>>);
+
+/// Error type for invalid endpoints.
+#[derive(Debug, Snafu)]
+#[snafu(context(suffix(false)))]
+pub enum EndpointError {
+    Parse { source: url::ParseError, domain: String },
+}
+
+#[derive(Default)]
+/// Holds the mapping between domains and their api keys.
+pub struct SingleDomainResolver {
+    domain: String,
+    api_keys: Vec<String>,
+}
+
+impl SingleDomainResolver {
+    /// The domain of the resolver.
+    pub fn domain(&self) -> &String {
+        &self.domain
+    }
+
+    /// The api keys associated with the domain.
+    pub fn api_keys(&self) -> &Vec<String> {
+        &self.api_keys
+    }
+
+    fn add_api_key(&mut self, api_key: String) {
+        self.api_keys.push(api_key);
+    }
+
+    fn set_domain(&mut self, domain: String) {
+        self.domain = domain;
+    }
+}
+
+/// Converts AdditionalEndpoints into a list of SingleDomainResolvers with its
+/// destination domain and api keys.
```

Comment:
Small nit: in Rust doc comments, since they're all Markdown-compatible, we try to surround type/function names in backticks so that they get rendered as nice monospaced inline code blocks.

In some cases, we also try to surround them in link brackets, since then rustdoc will autogenerate intra-doc links to let you navigate to the referenced types quickly.

```suggestion
/// Converts [`AdditionalEndpoints`] into a list of [`SingleDomainResolver`]s.
```

(Just a little extra context: you generally wouldn't need/want to do the link syntax if the doc comment is already on the type itself, or a method of that type... but since this is a free function, the link syntax is nicer because then we ensure that the person reading the docs in their IDE hover, or whatever, can quickly navigate to either of the referenced types.)